In [2]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder
from sklearn.metrics import accuracy_score
from sklearn.svm import SVC
import pandas as pd
import tensorflow as tf

# Import our input dataset
tele_df = pd.read_csv("/content/bank_telemarketing.csv")
tele_df.head()

Age       Job Marital_Status               Education Default_Credit  \
0   56     other        married       Primary_Education             no   
1   37  services        married     Secondary_Education             no   
2   40     admin        married       Primary_Education             no   
3   56  services        married     Secondary_Education             no   
4   59     admin        married  Professional_Education             no   

  Housing_Loan Personal_Loan Subscribed  
0           no            no         no  
1          yes            no         no  
2           no            no         no  
3           no           yes         no  
4           no            no         no

In [3]:
# Generate our categorical variable list
tele_cat = tele_df.dtypes[tele_df.dtypes == "object"].index.tolist()


# Check the number of unique values in each column
tele_df[tele_cat].nunique()

Job               9
Marital_Status    3
Education         4
Default_Credit    2
Housing_Loan      2
Personal_Loan     2
Subscribed        2
dtype: int64

In [6]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(sparse=False)

# Fit and transform the OneHotEncoder using the categorical variable list
encode_df = pd.DataFrame(enc.fit_transform(tele_df[tele_cat]))

# Add the encoded variable names to the dataframe
encode_df.columns = enc.get_feature_names(tele_cat)
encode_df.head()

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


Job_admin  Job_blue-collar  Job_entrepreneur  Job_management  Job_other  \
0        0.0              0.0               0.0             0.0        1.0   
1        0.0              0.0               0.0             0.0        0.0   
2        1.0              0.0               0.0             0.0        0.0   
3        0.0              0.0               0.0             0.0        0.0   
4        1.0              0.0               0.0             0.0        0.0   

   Job_retired  Job_self-employed  Job_services  Job_technician  \
0          0.0                0.0           0.0             0.0   
1          0.0                0.0           1.0             0.0   
2          0.0                0.0           0.0             0.0   
3          0.0                0.0           1.0             0.0   
4          0.0                0.0           0.0             0.0   

   Marital_Status_divorced  ...  Education_Secondary_Education  \
0                      0.0  ...                            0.0   
1                      0.0  ...                            1.0   
2                      0.0  ...                            0.0   
3                      0.0  ...                            1.0   
4                      0.0  ...                            0.0   

   Education_Tertiary_Education  Default_Credit_no  Default_Credit_yes  \
0                           0.0                1.0                 0.0   
1                           0.0                1.0                 0.0   
2                           0.0                1.0                 0.0   
3                           0.0                1.0                 0.0   
4                           0.0                1.0                 0.0   

   Housing_Loan_no  Housing_Loan_yes  Personal_Loan_no  Personal_Loan_yes  \
0              1.0               0.0               1.0                0.0   
1              0.0               1.0               1.0                0.0   
2              1.0               0.0               1.0                0.0   
3              1.0               0.0               0.0                1.0   
4              1.0               0.0               1.0                0.0   

   Subscribed_no  Subscribed_yes  
0            1.0             0.0  
1            1.0             0.0  
2            1.0             0.0  
3            1.0             0.0  
4            1.0             0.0  

[5 rows x 24 columns]

In [7]:
# Merge one-hot encoded features and drop the originals
tele_df = tele_df.merge(encode_df,left_index=True, right_index=True)
tele_df = tele_df.drop(tele_cat,1)
tele_df.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  This is separate from the ipykernel package so we can avoid doing imports until


Age  Job_admin  Job_blue-collar  Job_entrepreneur  Job_management  \
0   56        0.0              0.0               0.0             0.0   
1   37        0.0              0.0               0.0             0.0   
2   40        1.0              0.0               0.0             0.0   
3   56        0.0              0.0               0.0             0.0   
4   59        1.0              0.0               0.0             0.0   

   Job_other  Job_retired  Job_self-employed  Job_services  Job_technician  \
0        1.0          0.0                0.0           0.0             0.0   
1        0.0          0.0                0.0           1.0             0.0   
2        0.0          0.0                0.0           0.0             0.0   
3        0.0          0.0                0.0           1.0             0.0   
4        0.0          0.0                0.0           0.0             0.0   

   ...  Education_Secondary_Education  Education_Tertiary_Education  \
0  ...                            0.0                           0.0   
1  ...                            1.0                           0.0   
2  ...                            0.0                           0.0   
3  ...                            1.0                           0.0   
4  ...                            0.0                           0.0   

   Default_Credit_no  Default_Credit_yes  Housing_Loan_no  Housing_Loan_yes  \
0                1.0                 0.0              1.0               0.0   
1                1.0                 0.0              0.0               1.0   
2                1.0                 0.0              1.0               0.0   
3                1.0                 0.0              1.0               0.0   
4                1.0                 0.0              1.0               0.0   

   Personal_Loan_no  Personal_Loan_yes  Subscribed_no  Subscribed_yes  
0               1.0                0.0            1.0             0.0  
1               1.0                0.0            1.0             0.0  
2               1.0                0.0            1.0             0.0  
3               0.0                1.0            1.0             0.0  
4               1.0                0.0            1.0             0.0  

[5 rows x 25 columns]

In [8]:
# Remove Subscribed Status target from features data
y = tele_df.Subscribed_yes.values
X = tele_df.drop(columns=["Subscribed_no","Subscribed_yes"]).values

# Split training/test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [12]:
#Create the SVM model
svm = SVC(kernel='linear')
# Train the model
svm.fit(X_train, y_train)

SVC(kernel='linear')

In [13]:
# Evaluate the model
y_pred = svm.predict(X_test_scaled)
print(f" SVM model accuracy: {accuracy_score(y_test,y_pred):.3f}")


 SVM model accuracy: 0.873


In [14]:
# Define the model - deep neural net
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 10
hidden_nodes_layer2 = 5

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))


# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Compile the Sequential model together and customize metrics
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [15]:
# Train the model 

nn.fit(X_train_scaled,y_train , epochs=100)

# Evaluate the model using the test data 
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")


Epoch 1/100
715/715 [==============================] - 4s 4ms/step - loss: 0.3932 - accuracy: 0.8735
Epoch 2/100
715/715 [==============================] - 3s 3ms/step - loss: 0.3715 - accuracy: 0.8735
Epoch 3/100
715/715 [==============================] - 2s 3ms/step - loss: 0.3694 - accuracy: 0.8735
Epoch 4/100
715/715 [==============================] - 2s 3ms/step - loss: 0.3686 - accuracy: 0.8735
Epoch 5/100
715/715 [==============================] - 3s 4ms/step - loss: 0.3681 - accuracy: 0.8735
Epoch 6/100
715/715 [==============================] - 2s 2ms/step - loss: 0.3677 - accuracy: 0.8735
Epoch 7/100
715/715 [==============================] - 1s 2ms/step - loss: 0.3673 - accuracy: 0.8735
Epoch 8/100
715/715 [==============================] - 1s 2ms/step - loss: 0.3671 - accuracy: 0.8735
Epoch 9/100
715/715 [==============================] - 1s 2ms/step - loss: 0.3668 - accuracy: 0.8735
Epoch 10/100
715/715 [==============================] - 1s 2ms/step - loss: 0.3664 - accura

Looking at the results of our comparative analysis, the SVM and deep learning models both achieved a predictive accuracy around 87%. Additionally, both models take similar amounts of time to train on the input data. The only noticeable difference between the two models is implementation—the amount of code required to build and train the SVM is notably less than the comparable deep learning model. As a result, many data scientists will prefer to use SVMs by default, then turn to deep learning models, as needed.